### ISSUE TITLE
[Bug Report] hook_normalized / hook_scale interventions and backward hooks are silently ignored on NormalizationBridge's native-autograd path

### ISSUE THREAD

**Issue by Hetens** [TransformerLensOrg/TransformerLens#1526]
**Describe the bug**

When a `NormalizationBridge` is constructed with `use_native_layernorm_autograd=True`, its forward path (`_hf_autograd_forward_with_hooks` in `transformer_lens/model_bridge/generalized_components/normalization.py`) computes the values for `hook_scale` and `hook_normalized` inside a `torch.no_grad()` block and **discards the hooks' return values**:

```python
with torch.no_grad():
    ...
    x_normalized = x_centered * inv_rms
_ = self.hook_scale(scale)
_ = self.hook_normalized(x_normalized)
...
result = self.original_component(x)
```

The actual output always comes from `self.original_component(x)`. Two consequences, both silent:

1. **Forward interventions are discarded.** A hook that edits `hook_normalized` or `hook_scale` fires normally but has zero effect on model output. Anything built on patching these points — including the standard "freeze LN scale" linearization used in attribution patching — returns unpatched results with no error or warning.
2. **Backward hooks never fire.** The hook tensors are created under `no_grad`, so gradient-based attribution through these hook points silently produces nothing.

Caching/observation is unaffected (`ActivationCache.apply_ln_to_stack` etc. work fine) — which makes the failure harder to notice: the hooks *look* live.

The flag is hardwired in 13 adapters under `supported_architectures/`: `bart`, `bert`, `gemma3`, `gemma3_multimodal`, `gpt_oss`, `hubert`, `neox` (Pythia), `olmo`, `opt`, `phi`, `stablelm`, `t5gemma2`, `xglm`. On the default python-norm path (all other architectures) interventions and backward hooks work correctly, so the same user code behaves differently depending on which model is loaded.

**Code example**

Minimal repro against `dev` (`a041da3a`), CPU-only, no model download. The control (default path) passes; the native-autograd path fails both checks:

```python
import torch
import torch.nn as nn
from transformer_lens.model_bridge.generalized_components.normalization import NormalizationBridge

class Cfg:
    eps = 1e-5
    layer_norm_folding = False

def run(native: bool):
    torch.manual_seed(0)
    ln = nn.LayerNorm(8)
    bridge = NormalizationBridge(name="ln", config=Cfg(), use_native_layernorm_autograd=native)
    bridge.set_original_component(ln)
    x = torch.randn(2, 3, 8, requires_grad=True)

    baseline = bridge(x)
    bridge.hook_normalized.add_hook(lambda t, hook=None: torch.zeros_like(t))
    patched = bridge(x)
    bridge.hook_normalized.remove_hooks()
    edit_propagates = not torch.allclose(baseline, patched)

    bwd_fired = {}
    bridge.hook_normalized.add_hook(lambda g, hook=None: bwd_fired.setdefault("y", True) and g, dir="bwd")
    bridge(x).sum().backward()
    bridge.hook_normalized.remove_hooks()
    print(f"native={native}: edit propagates={edit_propagates}, bwd hook fired={'y' in bwd_fired}")

run(native=False)  # control:  edit propagates=True,  bwd hook fired=True
run(native=True)   # bug:      edit propagates=False, bwd hook fired=False
```

Output:

```
native=False: edit propagates=True, bwd hook fired=True
native=True: edit propagates=False, bwd hook fired=False
```

**System Info**

* Installed from source (`dev` branch, `a041da3a`)
* Linux, Python 3.11, torch 2.10.0, transformers 5.13.0, CPU

**Additional context**

The native path was introduced in #1101 for GPT-OSS dtype correctness, and its docstring is explicit that it exists "to preserve exact gradients" — so the observation-only design is understandable. The problem is that it is silent and undocumented: existing adapter tests only assert the flag is set, and no test covers intervention semantics on this path.

Possible directions, in increasing order of ambition:

1. **Warn or raise** when a hook attached to `hook_scale`/`hook_normalized` returns a modified value (or when a backward hook is attached) on this path — cheap, honest, stops silent wrong results.
2. **Fall back to the python-norm path when hooks are attached** to these points — preserves intervention semantics at the cost of slightly different numerics only while hooks are present.
3. Rework the native path to consume hook outputs (e.g. reconstruct the output from the hooked `x_normalized` and the module's weight/bias) — full fidelity, but touches the dtype/gradient guarantees the path exists for.

I'm happy to take this: I'd start with (1) plus a regression test that fails on current `dev` (checked via a clean worktree), and would like maintainer input on whether (2)/(3) are worth pursuing.

Two related minor observations in the same file (can be a separate PR if preferred):
- `_last_input_before_norm` (line 80) is written on every forward but never read anywhere in the repo since #1129 — it pins one `[batch, seq, d_model]` activation (and, when `requires_grad`, its upstream graph) per norm module between forwards.
- The `layer_norm_folding` config flag consumed at line 83 is never set `True` anywhere in the repo, so that branch appears unreachable in practice.

*This bug was found and this report drafted with the assistance of Claude (Anthropic's AI assistant), working under my direction; I've reviewed the analysis and the repro output is from a real run on my machine.*

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/TransformerLensOrg/TransformerLens/issues) in the repo (**required**)

**Comment by jlarson4** [comment]
Hi @Hetens! I was able to reproduce it and both side observations check out (`_last_input_before_norm` is never read + `layer_norm_folding` never set `True`). Please go with your suggested solutions 1 & 2, a warn/raise stopgap plus a regression test that's red on dev, then conditional dispatch that keeps the native path by default and only falls back to the python path when an editing forward hook or any backward hook is live on these points. One thing to watch out for, `run_with_cache` attaches observation-only hooks, so route on edit/bwd specifically rather than "any hook attached," or you'll change cached numerics for these 13 models. Feel free to submit the dead-code cleanup as a separate PR to keep the bug fix clean. 

Please base your fix on `main`, I'll make a quick patch release once this lands. Thank you for catching this bug!

**Comment by Hetens** [comment]
Thanks for the quick triage! Will do — one PR off `main` with the warn + conditional dispatch (routing only on detected edits / attached backward hooks, so `run_with_cache`'s observation-only hooks keep native numerics), plus red-on-unfixed regression tests. Dead-code cleanup will follow as a separate PR.
